x-pack/apm-server/sampling/eventstorage/storage.go

```diff
@@ -5,341 +5,80 @@
 package eventstorage
 
 import (
-	"bytes"
-	"errors"
-	"fmt"
-	"sync/atomic"
-	"time"
+	"io"
 
-	"github.com/dgraph-io/badger/v2"
+	"github.com/cockroachdb/pebble/v2"
 
 	"github.com/elastic/apm-data/model/modelpb"
 )
 
-const (
-	// NOTE(axw) these values (and their meanings) must remain stable
-	// over time, to avoid misinterpreting historical data.
-	entryMetaTraceSampled   = 's'
-	entryMetaTraceUnsampled = 'u'
-	entryMetaTraceEvent     = 'e'
-
-	// Initial transaction size
-	// len(txnKey) + 10
-	baseTransactionSize = 10 + 11
-)
-
-var (
-	// ErrNotFound is returned by by the Storage.IsTraceSampled method,
-	// for non-existing trace IDs.
-	ErrNotFound = errors.New("key not found")
-
-	// ErrLimitReached is returned by the ReadWriter.Flush method when
-	// the configured StorageLimiter.Limit is true.
-	ErrLimitReached = errors.New("configured storage limit reached")
-)
-
 type db interface {
-	NewTransaction(update bool) *badger.Txn
-	Size() (lsm, vlog int64)
-	Close() error
-}
-
-// Storage provides storage for sampled transactions and spans,
-// and for recording trace sampling decisions.
-type Storage struct {
-	db db
-	// pendingSize tracks the total size of pending writes across ReadWriters
-	pendingSize *atomic.Int64
-	codec       Codec
-}
-
-// Codec provides methods for encoding and decoding events.
-type Codec interface {
-	DecodeEvent([]byte, *modelpb.APMEvent) error
-	EncodeEvent(*modelpb.APMEvent) ([]byte, error)
-}
-
-// New returns a new Storage using db and codec.
-func New(db db, codec Codec) *Storage {
-	return &Storage{db: db, pendingSize: &atomic.Int64{}, codec: codec}
-}
-
-// NewShardedReadWriter returns a new ShardedReadWriter, for sharded
-// reading and writing.
-//
-// The returned ShardedReadWriter must be closed when it is no longer
-// needed.
-func (s *Storage) NewShardedReadWriter() *ShardedReadWriter {
-	return newShardedReadWriter(s)
-}
-
-// NewReadWriter returns a new ReadWriter for reading events from and
-// writing events to storage.
-//
-// The returned ReadWriter must be closed when it is no longer needed.
-func (s *Storage) NewReadWriter() *ReadWriter {
-	s.pendingSize.Add(baseTransactionSize)
-	return &ReadWriter{
-		s:           s,
-		txn:         nil, // lazy init to avoid deadlock in storage manager
-		pendingSize: baseTransactionSize,
-	}
+	Get(key []byte) ([]byte, io.Closer, error)
+	Set(key, value []byte, opts *pebble.WriteOptions) error
+	Delete(key []byte, opts *pebble.WriteOptions) error
+	NewIter(o *pebble.IterOptions) (*pebble.Iterator, error)
```

Comment:
[to reviewer] I've got rid of ShardedReadWriter, as I'm not using any transactions ("batch" in pebble). The removal of transaction actually makes a lot of things much simpler, and reduced memory usage. There's no more locking, and storage limit handling is much easier without sharded and in-flight transactions. I haven't noticed any downsides to it yet, but let me know if I missed anything